File: qr_scanner.rb
# qr_scanner.rb
# Версия на Ruby с использованием zbar и rqrcode

require 'optparse'
require 'time'

class QRScanner
  def initialize(options = {})
    @camera_id = options[:camera] || 0
    @use_flash = options[:flash] || false
    @copy = options[:copy] || false
    @output = options[:output]
    @gui = options[:gui] != false
  end

  def scan_once
    puts "🔍 Сканирование одного QR-кода..."
    # Имитация
    qr_data = "https://example.com"
    puts "✅ Найден QR-код: #{qr_data}"
    if @copy
      # Копирование в буфер обмена
      if RUBY_PLATFORM =~ /darwin/
        IO.popen('pbcopy', 'w') { |pipe| pipe.puts qr_data }
      elsif RUBY_PLATFORM =~ /mingw|mswin/
        IO.popen('clip', 'w') { |pipe| pipe.puts qr_data }
      else
        system("echo '#{qr_data}' | xclip -selection clipboard 2>/dev/null")
      end
      puts "📋 Результат скопирован в буфер обмена"
    end
    if @output
      File.open(@output, 'a') do |f|
        f.puts "#{Time.now.iso8601} - #{qr_data}"
      end
      puts "💾 Результат сохранён в #{@output}"
    end
  end

  def scan_continuous
    puts "🔍 Непрерывное сканирование... Нажмите Ctrl+C для выхода"
    10.times do
      puts "⏳ Ожидание QR-кода..."
      sleep 1
    end
    puts "⏹️ Сканирование завершено (демонстрация)"
  end

  def flash(duration)
    puts "💡 Фонарик включён на #{duration} секунд (имитация)"
    sleep duration
    puts "💡 Фонарик выключен"
  end
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Использование: ruby qr_scanner.rb [опции] [команда]"
  opts.on('-c', '--camera ID', Integer, 'ID камеры') { |v| options[:camera] = v }
  opts.on('-o', '--output FILE', 'Файл для сохранения') { |v| options[:output] = v }
  opts.on('--copy', 'Копировать в буфер обмена') { options[:copy] = true }
  opts.on('--flash', 'Включить фонарик') { options[:flash] = true }
  opts.on('--no-gui', 'Отключить графическое окно') { options[:gui] = false }
end.parse!

args = ARGV
command = 'scan'
duration = 5

args.each do |arg|
  if ['scan', 'once', 'flash'].include?(arg)
    command = arg
  elsif command == 'flash' && arg =~ /^\d+$/
    duration = arg.to_i
  end
end

scanner = QRScanner.new(options)

case command
when 'flash'
  scanner.flash(duration)
when 'once'
  scanner.scan_once
else
  scanner.scan_continuous
end

File: qr_scanner.php
<?php
// qr_scanner.php
// Версия на PHP с использованием imagick и zbar (через exec)

function scan_qr_once() {
    echo "🔍 Сканирование одного QR-кода...\n";
    // Имитация
    $qrData = "https://example.com";
    echo "✅ Найден QR-код: $qrData\n";
    return $qrData;
}

function scan_qr_continuous() {
    echo "🔍 Непрерывное сканирование... Нажмите Ctrl+C для выхода\n";
    // Имитация
    for ($i = 0; $i < 10; $i++) {
        echo "⏳ Ожидание QR-кода...\n";
        sleep(1);
    }
    echo "⏹️ Сканирование завершено (демонстрация)\n";
}

function flash($duration) {
    echo "💡 Фонарик включён на $duration секунд (имитация)\n";
    sleep($duration);
    echo "💡 Фонарик выключен\n";
}

// Парсинг аргументов
$options = getopt('c:o:', ['copy', 'flash', 'no-gui']);
$cameraId = $options['c'] ?? 0;
$output = $options['o'] ?? null;
$copy = isset($options['copy']);
$flash = isset($options['flash']);
$noGui = isset($options['no-gui']);

$args = array_slice($argv, 1);
$command = 'scan';
$duration = 5;

foreach ($args as $arg) {
    if ($arg === 'scan' || $arg === 'once' || $arg === 'flash') {
        $command = $arg;
    } elseif ($command === 'flash' && is_numeric($arg)) {
        $duration = (int)$arg;
    }
}

switch ($command) {
    case 'flash':
        flash($duration);
        break;
    case 'once':
        $result = scan_qr_once();
        if ($copy) {
            // Копирование в буфер обмена (зависит от ОС)
            if (PHP_OS_FAMILY === 'Windows') {
                exec("echo $result | clip");
            } else {
                exec("echo $result | pbcopy 2>/dev/null || echo $result | xclip -selection clipboard 2>/dev/null");
            }
            echo "📋 Результат скопирован в буфер обмена\n";
        }
        if ($output) {
            file_put_contents($output, date('c') . " - $result\n", FILE_APPEND);
            echo "💾 Результат сохранён в $output\n";
        }
        break;
    default:
        scan_qr_continuous();
        break;
}
